app/liquid_tags/liquid_tag_base.rb

```diff
@@ -3,11 +3,56 @@ def self.script
     ""
   end
 
+  def validate_data(parsed_context)
+    validate_contexts(parsed_context) if context_restricted?
+    validate_user_roles(parsed_context) if role_restricted?
+  end
+
   def finalize_html(input)
     input.gsub(/ {2,}/, "").
       gsub(/\n/m, " ").
       gsub(/>\n{1,}</m, "><").
       strip.
       html_safe
   end
+
+  private
+
+  def context_restricted?
+    self.class.const_defined? "VALID_CONTEXTS"
+  end
+
+  def role_restricted?
+    self.class.const_defined? "VALID_ROLES"
+  end
+
+  def validate_contexts(parsed_context)
+    source = parsed_context.partial_options[:source]
+    raise LiquidTags::Errors::InvalidParsedContext, "No source found" unless source
+
+    is_valid_source = self.class::VALID_CONTEXTS.include? source.class.name
+    invalid_source_error_msg = "Invalid context. This liquid tag can only be used in #{self.class::VALID_CONTEXTS.join(', ')}."
```

Comment:
Quick, non-blocking question: what do you think about changing this error message to be plural? 

The reason I ask is that when an error occurs, the message reads "Invalid context. This liquid tag can only be used in Article". By pluralizing it, when an error occurs, it will read a bit nicer: "Invalid context. This liquid tag can only be used in Articles" and it will make the error message reusable if and when we decide to add more valid contexts (as opposed to how it would less readable and reusable if we added an "an" to the error message and did not pluralize it).

Reply:
Updated!